### PR TITLE
netcdf: migrate from homebrew/science

### PR DESCRIPTION
**Pull request by iMichka**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by iMichka** [comment]
@ilovezfs I made the fortran option non-optional. I made some tests and it seems to be OK. It adds just some libraries more the bottle, so it should not break anything that does not need fortran. I think the errors we saw during the last updates were probably due to some formula that are broken anyway. So we should try this with fortran included: if stuff breaks in science, we will need to fix these formulas separately.

**Comment by ilovezfs** [comment]
@iMichka we already tried that.

https://github.com/Homebrew/homebrew-science/pull/4989

That time it broke gnudatalanguage. I have no reason to believe anything changed.

**Comment by iMichka** [comment]
I tested some builds on mac (but could not test everything), and it worked.

If the concern is gnudatalanguage on Linux, the error message was: `undefined reference to `Magick::Image::read` (and other equivalent messages). I already saw this somewhere on Linuxbrew, and it is probably unrelated. I think it is just some graphicsmagick linker flag missing somewhere.

I can retest this on my Linux box later today or tomorrow if you want.

**Comment by ilovezfs** [comment]
>I think it is just some graphicsmagick linker flag missing somewhere.

The CI went green again after we reverted, so it's almost certainly related.

**Comment by iMichka** [comment]
>  The CI went green again after we reverted, so it's almost certainly related.

Yes, but maybe indirectly. Don't know what happened here. Maybe the linker flag order is important, or some other thing. Let's wait that I am back and I'll test this to confirm before merging here.

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by iMichka** [comment]
Err this will work better if I put the file in the right folder :D

**Comment by fxcoudert** [comment]
Audit complains that:

```
  * Non-libraries were installed to "/usr/local/opt/netcdf/lib"
    Installing non-libraries to "lib" is discouraged.
    The offending files are:
      /usr/local/opt/netcdf/lib/libnetcdf-cxx.settings
      /usr/local/opt/netcdf/lib/libnetcdf.settings
```

What uses these files?

**Comment by fxcoudert** [comment]
I suggest we move the `libnetcdf{,-cxx}.settings` to `etc` or `libexec`.

**Comment by ilovezfs** [comment]
I would just ignore them. That audit is ridiculous.

**Comment by ilovezfs** [comment]
https://github.com/Homebrew/brew/pull/2997

**Comment by ilovezfs** [comment]
You can now safely ignore the audit failure, as it will only run for `--new-formula`

File: Formula/netcdf.rb [hunk 1]
```diff
@@ -0,0 +1,107 @@
+class Netcdf < Formula
+  desc "Libraries and data formats for array-oriented scientific data"
+  homepage "https://www.unidata.ucar.edu/software/netcdf"
+  url "ftp://ftp.unidata.ucar.edu/pub/netcdf/netcdf-4.4.1.1.tar.gz"
+  mirror "https://www.gfd-dennou.org/library/netcdf/unidata-mirror/netcdf-4.4.1.1.tar.gz"
+  sha256 "4d44c6f4d02a8faf10ea619bfe1ba8224cd993024f4da12988c7465f663c8cae"
```
Comment (ilovezfs) [inline Formula/netcdf.rb]: revision needs to be >= 5 (probably 6 if we're going to change the fortran default)

**Comment by iMichka** [comment]
I'll update the revision (5 or 6), depending on the decision for fortran.

I made some tests and the build error is now understood for `gnudatalanguage` on Linux: #6053
We have workaround for this, so I think it is okay to try with fortran enabled. Is that OK for you?

**Comment by ilovezfs** [comment]
@iMichka :+1: 

also wondering if we should do the same for hdf5 …

**Comment by fxcoudert** [comment]
I'd advise having Fortran enabled by default for both netcdf and hdf5. This is heavily used in numerical and scientific computing.

**Comment by iMichka** [comment]
Yes I think that would be great too. I'll just run a small check before doing that, I want to see how it affects the resulting hdf5 setup. I'll make a PR for that, and once it is merged I will bump the revision here.

**Comment by fxcoudert** [comment]
Fortran test:

```
(testpath/"test.f90").write <<-EOS.undent
program test
  use netcdf
  integer :: ncid, varid, dimids(2)
  integer :: dat(2,2) = reshape([1, 2, 3, 4], [2, 2])
  call check( nf90_create("test.nc", NF90_CLOBBER, ncid) )
  call check( nf90_def_dim(ncid, "x", 2, dimids(2)) )
  call check( nf90_def_dim(ncid, "y", 2, dimids(1)) )
  call check( nf90_def_var(ncid, "data", NF90_INT, dimids, varid) )
  call check( nf90_enddef(ncid) )
  call check( nf90_put_var(ncid, varid, dat) )
  call check( nf90_close(ncid) )
contains
  subroutine check(status)
    integer, intent(in) :: status
    if (status /= nf90_noerr) call abort
  end subroutine check  
end program test
EOS
system ENV.fc, "test.f90", "-L#{lib}", "-I#{include}", "-lnetcdff", "-o", "testf"
system "./test"
```

File: Formula/netcdf.rb [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  end
+
+  def install
+    ENV.deparallelize
+
+    common_args = std_cmake_args << "-DBUILD_SHARED_LIBS=ON"
+    common_args << "-DBUILD_TESTING=OFF"
+
+    mkdir "build" do
+      # Intermittent availability of the DAP endpoints tests means that sometimes
```
Comment (fxcoudert) [inline Formula/netcdf.rb]: This comment can go away… we don't usually run tests anyway

**Comment by ilovezfs** [comment]
How are we feeling about this?

File: Formula/netcdf.rb [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        call check( nf90_put_var(ncid, varid, dat) )
+        call check( nf90_close(ncid) )
+      contains
+        subroutine check(status)
+          integer, intent(in) :: status
+          if (status /= nf90_noerr) call abort
+        end subroutine check
+      end program test
+      EOS
+    system ENV.fc, "test.f90", "-L#{lib}", "-I#{include}", "-lnetcdff", "-o", "testf"
```
Comment (fxcoudert) [inline Formula/netcdf.rb]: Forgot that it needs `ENV.fortran` before we call `ENV.fc`

**Comment by fxcoudert** [comment]
Other than the missing `ENV.fortran` (my bad), I'm 👍👍👍

**Comment by iMichka** [comment]
Good for me. #6028 will need to be merged as soon as this goes in. I'll merge core to linuxbrew as soon as possible. It will rebuild all reverse deps for science; so I'll have a look if there is any outstanding problem. I'll also open a PR for `gnudatalanguage` asap.

**Comment by fxcoudert** [comment]
Thanks @iMichka @ilovezfs!

**Pull request closed without merging**